### PR TITLE
github: Switch to dynamic filtering

### PR DESCRIPTION
**Pull request by msfjarvis**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :bulb: Motivation and Context
`paths` filter is incompatible with necessary checks system since it blocks merges when the workflow is not run, causing an endless stream of pain.

## :pencil: Checklist

- [ ] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [ ] I reviewed submitted code

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -19,7 +10,25 @@ jobs:
...
+      with:
+        result-encoding: string
+        script: |
+          const result = await github.pulls.listFiles({
+            owner: context.payload.repository.owner.login,
+            repo: context.payload.repository.name,
+            pull_number: context.payload.number,
+            per_page: 100
+          })
+          const serviceChanged = result.data.filter(f => f.filename.startsWith("app/") || f.filename.endsWith("gradle") || f.filename.startsWith(".github")).length > 0
```
Comment (FabianHenneke) [inline .github/workflows/pull_request.yml:25]: Do we also need a `startsWith("gradle")` and `endsWith("properties")` here?

Comment (msfjarvis) [reply .github/workflows/pull_request.yml:25]: Probably yeah, fixing.

**Comment by msfjarvis** [comment]
https://github.com/msfjarvis/Android-Password-Store/pull/1 steps get skipped when I only update README and changing gradle.properties correctly starts the full suite of checks

File: .github/workflows/pull_request.yml [hunk 2]
```diff
@@ -19,15 +10,36 @@ jobs:
...
+      with:
+        result-encoding: string
+        script: |
+          const result = await github.pulls.listFiles({
+            owner: context.payload.repository.owner.login,
+            repo: context.payload.repository.name,
+            pull_number: context.payload.number,
+            per_page: 100
+          })
+          const serviceChanged = result.data.filter(f => f.filename.startsWith("app/") || f.filename.endsWith("gradle") || f.filename.startsWith(".github") || (f.filename.startsWith("gradle") && f.filename.endsWith("properties"))).length > 0
```
Comment (FabianHenneke) [inline .github/workflows/pull_request.yml:25]: Don't we also need to run the checks when making changes to `gradlew`? Maybe the `&&` should just be an `||`.

Comment (msfjarvis) [reply .github/workflows/pull_request.yml]: Yeah that's right. Fixed it.

**Comment by msfjarvis** [comment]
Had to rebase without changes to fix the broken test

**Pull request merged**